### PR TITLE
[Resource] Add new parameters to the What-If cmdlets to support excluding change types

### PR DESCRIPTION
**Pull request by shenglol**
## Description

* Added `-ExcludeChangeTypes` parameter to `Get-AzDeploymentWhatIfResult` and `Get-AzResourceGroupDeploymentWhatIfResult`
* Added `-WhatIfExcludeChangeTypes` parameter to `New-AzDeployment` and `New-AzResourceGroupDeployment`
* Updated `Test-Az*Deployment` cmdlets to show better error messages

This addresses https://github.com/Azure/arm-template-whatif/issues/53.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentWhatIfResultCmdlet.cs [hunk 1]
```diff
@@ -45,6 +50,12 @@ public class GetAzureSubscriptionDeploymentWhatIfResultCmdlet : ResourceWithPara
         [Parameter(Mandatory = false, HelpMessage = "The What-If result format.")]
         public WhatIfResultFormat ResultFormat { get; set; } = WhatIfResultFormat.FullResourcePayloads;
 
+        [Parameter(Mandatory = false, HelpMessage = "Comma-separated list of resource change types to be excluded from What-If results.")]
+        [ChangeTypeCompleter]
+        [ValidateChangeTypes]
+        public string[] ExcludeChangeTypes { get; set; }
```
Comment (VeryEarly) [inline src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentWhatIfResultCmdlet.cs:56]: Please change name to ExcludeChangeType

File: src/Resources/ResourceManager/Implementation/Deployments/NewAzureSubscriptionDeploymentCmdlet.cs [hunk 2]
```diff
@@ -54,9 +55,14 @@ public class NewAzureSubscriptionDeploymentCmdlet : ResourceWithParameterCmdletB
...
         public Hashtable Tag { get; set; }
         
-        [Parameter(Mandatory = false, HelpMessage = "The What-If result format.")]
+        [Parameter(Mandatory = false, HelpMessage = "The What-If result format. Applicable when the -WhatIf or -Confirm switch is set.")]
         public WhatIfResultFormat WhatIfResultFormat { get; set; } = WhatIfResultFormat.FullResourcePayloads;
 
+        [Parameter(Mandatory = false, HelpMessage = "Comma-separated resource change types to be excluded from What-If results. Applicable when the -WhatIf or -Confirm switch is set.")]
+        [ChangeTypeCompleter]
+        [ValidateChangeTypes]
+        public string[] WhatIfExcludeChangeTypes { get; set; }
```
Comment (VeryEarly) [inline src/Resources/ResourceManager/Implementation/Deployments/NewAzureSubscriptionDeploymentCmdlet.cs:64]: please change parameter name to WhatIfExcludeChangeType

File: src/Resources/ResourceManager/Implementation/ResourceGroupDeployments/GetAzureResourceGroupDeploymentWhatIfResultCmdlet.cs [hunk 3]
```diff
@@ -44,6 +45,11 @@ public class GetAzureResourceGroupDeploymentWhatIfResultCmdlet : ResourceWithPar
         [Parameter(Mandatory = false, HelpMessage = "The What-If result format.")]
         public WhatIfResultFormat ResultFormat { get; set; } = WhatIfResultFormat.FullResourcePayloads;
 
+        [Parameter(Mandatory = false, HelpMessage = "Comma-separated resource change types to be excluded from What-If results.")]
+        [ChangeTypeCompleter]
+        [ValidateChangeTypes]
+        public string[] ExcludeChangeTypes { get; set; }
```
Comment (VeryEarly) [inline src/Resources/ResourceManager/Implementation/ResourceGroupDeployments/GetAzureResourceGroupDeploymentWhatIfResultCmdlet.cs:51]: please change parameter name to ExcludeChangeType

File: src/Resources/ResourceManager/Implementation/ResourceGroupDeployments/NewAzureResourceGroupDeploymentCmdlet.cs [hunk 4]
```diff
@@ -60,9 +61,14 @@ public class NewAzureResourceGroupDeploymentCmdlet : ResourceWithParameterCmdlet
...
         public Hashtable Tag { get; set; }
         
-        [Parameter(Mandatory = false, HelpMessage = "The What-If result format.")]
+        [Parameter(Mandatory = false, HelpMessage = "The What-If result format. Applicable when the -WhatIf or -Confirm switch is set.")]
         public WhatIfResultFormat WhatIfResultFormat { get; set; } = WhatIfResultFormat.FullResourcePayloads;
 
+        [Parameter(Mandatory = false, HelpMessage = "Comma-separated resource change types to be excluded from What-If results. Applicable when the -WhatIf or -Confirm switch is set.")]
+        [ChangeTypeCompleter]
+        [ValidateChangeTypes]
+        public string[] WhatIfExcludeChangeTypes { get; set; }
```
Comment (VeryEarly) [inline src/Resources/ResourceManager/Implementation/ResourceGroupDeployments/NewAzureResourceGroupDeploymentCmdlet.cs:70]: please change parameter name to WhatIfExcludeChangeType

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentWhatIfResultCmdlet.cs (hunk 1 above)
Comment (shenglol) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureSubscriptionDeploymentWhatIfResultCmdlet.cs]: Thanks. Changed it to singular.

**Comment by shenglol** [comment]
> please check comments,
> is it possible to have test cases cover your code changes?

Yes. Added tests to cover the new parameter and the formatter.

**Pull request merged**